packages/dds/merge-tree/docs/Obliterate.md

```diff
@@ -0,0 +1,639 @@
+
+This document covers motivation, spec, and design for the upcoming "obliterate" feature of merge-tree.
+
+## Spec
+
+A concise description of merge-tree's current merge conflict resolution strategy is as follows:
+
+- Insertion of a text segment only conflicts with other insertions at the same location.
+  The conflict is resolved by inserting the segment added later nearer in the string.
+  For example, from an initial state of "abc", if the operations [insert "hi " at 0] from client 1
+  and [insert "bye " at 0] from client 2 are sequenced in that order, the resulting state is "bye hi abc".
+- Range operations (delete, annotate) apply to the range at the time the operation was issued.
+  Specifically, insertion of a segment into a range that is concurrently deleted or annotated
+  will not result in that inserted segment being deleted or annotated. For example, from an initial state "012",
+  the operations [delete the range [1, 3)] from client 1 and [insert "hi" at index 2 (i.e. between "1" and "2")] from client 2,
+  the resulting text is "0hi".
+
+The merge outcomes for ranges are easy to understand, but not always desirable.
+Oftentimes, when consumers want to work with ranges, they may want their operation to apply to concurrently inserted segments.
+In the example above, these semantics would look like so:
+
+```
+// Initial state at seq 0: "012"
+{ seq: 1, refSeq: 0, clientId: 1, op: <insert "hi" at index 2> }
+{ seq: 2, refSeq: 0, clientId: 2, op: <delete the range [1, 3)> }
+// final desired state: "0"
+```
+
+```
+// Initial state at seq 0: "012"
+{ seq: 1, refSeq: 0, clientId: 2, op: <delete the range [1, 3)> }
+{ seq: 2, refSeq: 0, clientId: 1, op: <insert "hi" at index 2> }
+// final desired state: "0"
+```
+
+A `SharedString` feature request for a removal operation with these semantics dubbed them "obliterate".
+
+At an implementation level, these semantics can be viewed in two parts:
+- The range specification is resolved at the time the op is sequenced
+- Any subsequent segments inserted into that range concurrently should also be removed
+
+The first clause handles concurrent inserts before the removal is sequenced, and the second clause handles concurrent inserts after the removal is sequenced.
+
+However, there is a way to view obliterate's semantics as a special case of a "move" operation,
+which preserves content identity such that concurrently inserted segments will be inserted to the range at its current location.
+A main motivator here from the app perspective might be the idea that if user 1 cut and pastes an entire paragraph to a different section of the document
+whiler user 2 edits it, the desired merge outcome would likely be for user 2's edit to apply to the paragraph in its new location.
+Roughly, anywhere an application would want obliterate merge semantics on user delete of some content,
+the same application would want move semantics if the user instead cut and pasted the content somewhere else.
```

Comment:
back in the day there was a feature in the merge tree we called registers, basically it allowed a client to specify a range, and put that range in a named register with an op, and then re-insert that range with an op by specifying the register. we removed the feature as it had problems around summarization, the potential for abuse, and wasn't really eventually consistent (incoming ops in the range had undefined behavior). It interesting in this context as to really implement something like cut and then paste we'd need a place to stash that content as it would probably be two different ops, one to cut, then one to paste. I wonder if that cut would be obliterate, or a special form of obliterate that would then support re-insertion. We'd still need to solve the summary and abuse potential issues, generally we probably want to only allow one such preserved content region per client session (probably want it to survive reconnection), and have it go away on session end.  I don't think this changes anything, but though it was some useful context when thinking about this space.

Reply:
the meta point here is, is move always a single operation, or should it be modeled to support a two phase approach, obliterate, then move/re-insert.